motoko/passwords/README.md

```diff
@@ -0,0 +1,78 @@
+# Design Pattern: Usernames & Passwords
+
+![Compatibility](https://img.shields.io/badge/compatibility-0.6.24-blue)
+[![Build Status](https://github.com/dfinity/examples/workflows/motoko-password-example/badge.svg)](https://github.com/dfinity/examples/actions?query=workflow%3Amotoko-password-example)
+
+This example demonstrates how a more traditional username and password authentication scheme can be used alongside the `caller` functionality within a Motoko actor.
+
+## Overview
+
+End-users of an application might want to access their account from multiple devices, a friend's device, or a shared device like at a library. While the Internet Computer provides authentication with every request in the form of the `caller` principal identifier, it can be useful to associate that principal with an existing account through the common pattern of a username and password.
+
+## Implementation
+
+The back-end canister exposes several methods for the lifecycle of a user account. A new user can `signup` with a specified username and password (which are then stored in a secure way). An existing user can `login` to associate a new device to an existing account or `logout` to remove that association. An authenticated user can ask `whoami` to get the currently authenticated account.
+
+Note: There are many obvious improvements (storing accounts in a more optimal data structure, implementing a real hashing algorithm) but hopefully this example illustrates the concepts in a simple way.
+
+## Prerequisites
+
+Verify the following before running this demo:
+
+*  You have downloaded and installed the [DFINITY Canister
+   SDK](https://sdk.dfinity.org).
+
+*  You have stopped any Internet Computer or other network process that would
+   create a port conflict on 8000.
+
+*  You have installed the front-end dependencies by running `npm install`.
+
+## Demo
+
+1. Start a local internet computer.
+
+   ```text
+   dfx start
+   ```
+
+1. Open a new terminal window.
+
+1. Create, build, and install both canisters.
+
+   ```text
+   dfx deploy
+   ```
+
+1. Create a new account.
+
+   ```text
+   dfx canister call passwords signup 'record { username = "Alice"; password = "p4ssw0rd" }'
+   ```
+
+1. Authenticate your command line identity.
+
+   ```text
+   dfx canister call passwords login 'record { username = "Alice"; password = "p4ssw0rd" };
```

Comment:
It might be a good idea to put a big warning here saying "DO NOT USE A PASSWORD YOU USE ON OTHER SITES". (It's my understanding we currently really make no promises about this value staying confidential from people who shouldn't have it. Dom might have a plan, but IMO he hasn't made a convincing argument that it's safe to tell the IC real secrets).

Even better, dont' take the risk at all. Do the password hashing (ideally [salted](https://en.wikipedia.org/wiki/Salt_(cryptography))... if we had a `serviceId` that might be a good salt) here in the shell.

Instead of storing `password`, store something like `password_hash` or, better, can even support a few password hash schemes, so like `password_hash_bcrypt` https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#password-hashing-algorithms

A naive hashing scheme would just be md5, and if it were, this line could be replaced with:

```
password_hash_md5=$(echo-my-password-to-stdout | md5)
dfx canister call passwords login 'record { username = "Alice"; password_hash_md5 = "$password_hash_md5" };
```

Now my password never was sent to IC.

There are probably more things in OWASP worth checking this against. I only looked it up to check if `bcrypt` is still a good one. https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#introduction